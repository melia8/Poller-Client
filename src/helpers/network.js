export async function fetchData(url) {
  try {
    const response = await fetch(url);
    const jsonData = await response.json();
    return jsonData;
  } catch (error) {
    console.error("Error fetching data:", error);
  }
}

export async function sendData(url, body) {
  try {
    const response = await fetch(url, {
      method: "POST",
      body: JSON.stringify(body),
      headers: {
        "Content-Type": "application/json",
      },
    });
    const jsonData = await response.json();
    return jsonData;
  } catch (error) {
    console.error("Error fetching data:", error);
  }
}
