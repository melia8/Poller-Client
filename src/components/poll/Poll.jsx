import PollBar from "./PollBar.jsx";
import classes from "./Poll.module.css";
import { useEffect, useState } from "react";
import { fetchData } from "../../helpers/network.js";
import Header from "../header/Header.jsx";

const livePollUrl = "http://localhost:8080/poll";

function Poll() {
  const [pollData, setPollData] = useState({});

  useEffect(() => {
    fetchData(livePollUrl).then((data) => {
      setPollData(data);
    });
  }, []);

  return (
    <div className={classes.layout}>
      <Header title={pollData.title} />
      {pollData.options &&
        pollData.options.map((data) => (
          <PollBar
            key={data.id}
            value={data.total}
            total={pollData.total}
            text={data.name}
          />
        ))}
    </div>
  );
}

export default Poll;
