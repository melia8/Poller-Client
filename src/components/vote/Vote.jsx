import { fetchData, sendData } from "../../helpers/network";
import classes from "./Vote.module.css";
import VoteBar from "./VoteBar";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import Header from "../header/Header";

const voteUrl = "http://localhost:8080/vote";
const livePollUrl = "http://localhost:8080/poll";

function Vote() {
  const [pollData, setpollData] = useState({});
  const [selected, setSelected] = useState("");
  const navigate = useNavigate();

  useEffect(() => {
    fetchData(livePollUrl).then((data) => {
      setpollData(data);
    });
  }, []);

  const updatePoll = (id, name) => {
    const vote = {
      pollId: id,
      name: name,
    };
    sendData(voteUrl, vote);
  };

  const submit = () => {
    updatePoll(pollData.id, selected);
    navigate("/poll");
  };

  return (
    <div className={classes.layout}>
      <Header title={pollData.title} />
      {pollData.options &&
        pollData.options.map((data) => (
          <VoteBar
            key={data.id}
            text={data.name}
            clicked={data.name === selected}
            select={() => setSelected(data.name)}
          />
        ))}

      <div
        disabled={selected ? false : true}
        onClick={submit}
        className={classes.submitBtn}
      >
        <span className={classes.submitText}>Submit</span>
      </div>
    </div>
  );
}

export default Vote;
