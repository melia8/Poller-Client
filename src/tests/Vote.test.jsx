import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import Poll from "../components/poll/Poll";
import Vote from "../components/vote/Vote";
import { sendData, fetchData } from "../helpers/network";
import { act } from "react";
import getTestData from "./testData";

vi.mock("../helpers/network.js", () => ({
  fetchData: vi.fn(),
  sendData: vi.fn(),
}));

fetchData.mockImplementation(async () => {
  return getTestData();
});

const renderVote = () => {
  return render(
    <MemoryRouter initialEntries={["/"]}>
      <Routes>
        <Route index element={<Vote />} />
        <Route path="/poll" element={<Poll />} />
      </Routes>
    </MemoryRouter>,
  );
};

describe("Vote", () => {
  it("The component UI should not change", async () => {
    expect(await act(async () => renderVote())).toMatchSnapshot();
  });

  it("renders correct teams", async () => {
    await act(async () => renderVote());
    screen.getByText("Liverpool");
    screen.getByText("Manchester City");
    screen.getByText("Arsenal");
    screen.getByText("Liverpool");
    screen.getByText("West Ham United");
  });

  it("sends correct vote", async () => {
    await act(async () => renderVote());

    await act(async () => {
      screen.getByText("Liverpool").click();
    });

    await act(async () => {
      screen.getByText("Submit").click();
    });

    expect(sendData).toBeCalledTimes(1);
    expect(sendData).toHaveBeenCalledWith("http://localhost:8080/vote", {
      pollId: 1,
      name: "Liverpool",
    });
  });
});
