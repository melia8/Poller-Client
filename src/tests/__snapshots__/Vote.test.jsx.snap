// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vote > The component UI should not change 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_layout_d6e136"
      >
        <div
          class="_logo_69e643"
        >
          <a
            class="_logoLink_69e643"
            href="https://dizplai.com/home/"
            rel="nofollow"
          >
            <img
              alt="Logo"
              height="60"
              src="https://dizplai.com/wp-content/themes/neverno/library/images/logo.svg"
              width="124"
            />
          </a>
        </div>
        <div
          class="_title_69e643"
        >
           
          Who will win Premier League
        </div>
        <div
          class="_voteBarBackground_3a229c"
        >
          <div
            class="_title_3a229c"
          >
             
            Liverpool
             
          </div>
        </div>
        <div
          class="_voteBarBackground_3a229c"
        >
          <div
            class="_title_3a229c"
          >
             
            Manchester City
             
          </div>
        </div>
        <div
          class="_voteBarBackground_3a229c"
        >
          <div
            class="_title_3a229c"
          >
             
            Arsenal
             
          </div>
        </div>
        <div
          class="_voteBarBackground_3a229c"
        >
          <div
            class="_title_3a229c"
          >
             
            Manchester United
             
          </div>
        </div>
        <div
          class="_voteBarBackground_3a229c"
        >
          <div
            class="_title_3a229c"
          >
             
            West Ham United
             
          </div>
        </div>
        <div
          class="_submitBtn_d6e136"
          disabled=""
        >
          <span
            class="_submitText_d6e136"
          >
            Submit
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_layout_d6e136"
    >
      <div
        class="_logo_69e643"
      >
        <a
          class="_logoLink_69e643"
          href="https://dizplai.com/home/"
          rel="nofollow"
        >
          <img
            alt="Logo"
            height="60"
            src="https://dizplai.com/wp-content/themes/neverno/library/images/logo.svg"
            width="124"
          />
        </a>
      </div>
      <div
        class="_title_69e643"
      >
         
        Who will win Premier League
      </div>
      <div
        class="_voteBarBackground_3a229c"
      >
        <div
          class="_title_3a229c"
        >
           
          Liverpool
           
        </div>
      </div>
      <div
        class="_voteBarBackground_3a229c"
      >
        <div
          class="_title_3a229c"
        >
           
          Manchester City
           
        </div>
      </div>
      <div
        class="_voteBarBackground_3a229c"
      >
        <div
          class="_title_3a229c"
        >
           
          Arsenal
           
        </div>
      </div>
      <div
        class="_voteBarBackground_3a229c"
      >
        <div
          class="_title_3a229c"
        >
           
          Manchester United
           
        </div>
      </div>
      <div
        class="_voteBarBackground_3a229c"
      >
        <div
          class="_title_3a229c"
        >
           
          West Ham United
           
        </div>
      </div>
      <div
        class="_submitBtn_d6e136"
        disabled=""
      >
        <span
          class="_submitText_d6e136"
        >
          Submit
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
